packages/core/src/browser/frontend-application.ts

```diff
@@ -224,6 +239,38 @@ export class FrontendApplication {
         }
     }
 
+    /**
+     * `beforeunload` listener implementation
+     */
+    protected getBeforeUnloadHandler() {
+        return (event: BeforeUnloadEvent) => {
+            // tslint:disable-next-line:no-any
+            const preventExit = (action?: any): void => {
```

Comment:
Could it be?
```ts
for (const contribution of this.contributions.getContributions()) {
    if (contribution.onWillStop) {
        const returnValue = contribution.onWillStop(app);
        if (returnValue) {
            event.returnValue = returnValue;
            event.preventDefault();
            return;
        }
    }
}
```

And let each contribution decide which preference to check and whether to use user notifications?